src/Spago/Command/Fetch.purs

```diff
@@ -278,6 +240,75 @@ run { packages: packagesToInstall, ensureRanges, isTest } = do
+  { workspacePackages :: Map PackageName Lock.WorkspaceLockPackage
+  , packages :: Map PackageName Lock.LockEntry
+  }
+
+mkLockfile :: forall a. PackageTransitiveDeps -> Spago (FetchEnv a) Lock.Lockfile
+mkLockfile allTransitiveDeps = do
+  { workspace, logOptions, getMetadata } <- ask
+  let
+    processPackage :: LockfileBuilderResult -> Tuple PackageName (Tuple PackageName Package) -> Spago (FetchEnv a) LockfileBuilderResult
+    processPackage result (Tuple workspacePackageName (Tuple dependencyName dependencyPackage)) = do
+      (packageDependencies :: Array PackageName) <- (Array.fromFoldable <<< Map.keys <<< fromMaybe Map.empty)
+        <$> getPackageDependencies dependencyName dependencyPackage
+      let
+        updatePackage package = pure $ result
+          { packages = Map.alter
+              ( case _ of
+                  Nothing -> Just { needed_by: NEA.singleton workspacePackageName, package }
+                  Just { needed_by } -> Just { needed_by: NEA.cons workspacePackageName needed_by, package }
+              )
+              dependencyName
+              result.packages
+          }
+        updateWorkspacePackage otherWorkspacePackage = pure $ result
+          { workspacePackages = Map.alter
+              ( case _ of
+                  Nothing -> Just $ otherWorkspacePackage { needed_by = Array.singleton workspacePackageName }
+                  Just pkg@{ needed_by } -> Just $ pkg { needed_by = Array.cons workspacePackageName needed_by }
+              )
+              dependencyName
+              result.workspacePackages
+          }
+
+      case dependencyPackage of
+        WorkspacePackage pkg -> updateWorkspacePackage (snd $ Config.workspacePackageToLockfilePackage pkg)
+        GitPackage gitPackage -> do
+          let packageLocation = Config.getPackageLocation dependencyName dependencyPackage
+          Git.getRef (Just packageLocation) >>= case _ of
+            Left err -> die err -- TODO maybe not die here?
+            Right rev -> updatePackage $ FromGit { rev, dependencies: packageDependencies, url: gitPackage.git, subdir: gitPackage.subdir }
+        RegistryVersion version -> do
+          metadata <- runSpago { logOptions } $ getMetadata dependencyName
+          registryVersion <- case (metadata >>= (\(Metadata meta) -> Either.note "Didn't find version in the metadata file" $ Map.lookup version meta.published)) of
+            Left err -> die $ "Couldn't read metadata, reason:\n  " <> err
+            Right { hash: integrity } ->
+              pure { version, integrity, dependencies: packageDependencies }
+          updatePackage $ FromRegistry registryVersion
+        LocalPackage { path } -> do
+          updatePackage $ FromPath { path, dependencies: packageDependencies }
+
+  let
+    toArray :: forall k v. Map k v -> Array (Tuple k v)
+    toArray = Map.toUnfoldable
+  ({ packages, workspacePackages } :: LockfileBuilderResult) <-
+    Array.foldM processPackage
```

Comment:
We used to just map the packages into the lockfile data structure, but we now use a fold to go through them so we can keep track of which workspace package needs which other package, which makes it dead easy to figure out the build plan afterwards